proposals/0540-jsem.rst

```diff
@@ -0,0 +1,384 @@
+
+Costs and Drawbacks
+-------------------
+
+The implementation in GHC is self-contained and doesn't impact the rest of the
+compiler much. It does however add a new flag (which interacts with ``-j``),
+and a complete implementation requires coordination with ``cabal`` and ``stack``.
+
+Alternatives
+------------
+
+GNU make supports a Jobserver protocol [9]_ [5]_ which is the same as the
+GHC Jobserver protocol described above, except that:
+
+- it uses POSIX pipes to exchange token's between processes.
+
+- participants in the protocol learn about it through environment variables
+  and the state of file descriptors on process entry.
+
+For example, rust's ``cargo`` implement the GNU make Jobserver protocol [13]_ .
+A prototype implementation of the GNU make Jobserver protocol for GHC was also
+made by Ellie Hermaszewska [15]_ .
+
+However, we have decided to depart from this design, for the following reasons:
+
+- Other communities have considered the Make jobserver, and decided that some
+  aspects of the protocol are unsuitable (OCaml [10]_ [11]_ , Nix [12]_ ,
+  ninja [16]_ ). To summarise:
+
+  - The protocol relies on spawned processes cooperating and returning tokens on
+    termination. If this doesn't happen, semaphore tokens can be lost entirely.
+    In comparison, with the approach described in this proposal, implicit tokens
+    are controlled by the server; this means that, at worst, the build will
+    continue with reduced parallelism, and mitigation strategies are available.
+
+  - The protocol uses anonymous file descriptors to communicate between
+    processes. This seems to be fragile, with many edge-cases. In 2019,
+    changes to the Linux kernel broke the Make jobserver protocol, due to
+    subtle changes in the semantics of pipes [18]_.
+
+- We expect ``cabal-install`` and ``stack`` to be the only users of this feature
+  in the near term. We think the proposed protocol is adequate for this use case.
+  ``-jsem`` doesn't provide a general solution either for mixed-language code bases
+  which require coordination with other build tools but then there isn't a widely
+  adopted solution which does.
+
+- We can extend GHC to use the GNU make Jobserver protocol in the future, if
+  there are users for it.
+
+Some operating systems also have OS-specific methods of mediating parallelism
+between processes. For example, MacOS's "Grand Central Dispatch" mechanism
+allows applications to queue up tasks to be run in parallel, and handles
+the scheduling.
+In order to implement parallelism at this level, it seems necessary to modify*
+the RTS and GHC's own thread scheduling algorithms. Not only this, the
+implementation would be specific to a platform.
+
+
+The other option for build systems is to use GHC in one-shot mode. In that case,
+the build system can control the scheduling of all the jobs. This is what
+``hadrian`` and ``rules_haskell`` do when building projects (``cabal``
+currently does not).
```

Comment:
This is what I wanted to ask about. Why not granularly split up the compilation by module and invoke GHC in one-shot mode? That would seem to substantially reduce the jobserver protocol complexity, and better make use of resources. I'd like to understand what the downsides are that you're seeing?

Reply:
Well you can do that, but it's not how cabal-install works and modifying cabal-install to work like that is a completely separate project which has no financial backing nor good time-bound for how long it would take. Using `--make` mode is also faster than using one-shot mode.